### ISSUE TITLE
Double reporting teardown failure

### ISSUE THREAD

**Issue by gruberma** [pytest-dev/pytest#9909]
When the teardown fails, the last test case gets reported twice, both with it's own verdict, and with the error of the teardown.
This also happens in the junitxml, but only if the last test case fails (example 2).
My main question is: is this behavior intended (especially the double reporting within the junit-xml where the same test case shows up twice)?

There is a [related issue](https://github.com/pytest-dev/pytest/issues/3850) reporting that the test number is wrong if the teardown fails, which also shows up here.

## Example 1: Last test case passes

```python
def test_fail():
    assert False

def test_pass():
    pass

def teardown():
    assert 1 == 0
```

```bash
python3 -m pytest -v --junit-xml junit.xml test_double_outcome.py
===================== test session starts =====================
platform linux -- Python 3.9.2, pytest-7.1.2, pluggy-1.0.0 -- REMOVED/venv/bin/python3
cachedir: .pytest_cache
rootdir: REMOVED
collected 2 items                                             

test_double_outcome.py::test_fail FAILED                [ 50%]
test_double_outcome.py::test_pass PASSED                [100%]
test_double_outcome.py::test_pass ERROR                 [100%]

=========================== ERRORS ============================
_______________ ERROR at teardown of test_pass ________________

    def teardown():
>       assert 1 == 0
E       assert 1 == 0

test_double_outcome.py:8: AssertionError
========================== FAILURES ===========================
__________________________ test_fail __________________________

    def test_fail():
>       assert False
E       assert False

test_double_outcome.py:2: AssertionError
- generated xml file: REMOVED/junit.xml -
=================== short test summary info ===================
FAILED test_double_outcome.py::test_fail - assert False
ERROR test_double_outcome.py::test_pass - assert 1 == 0
============ 1 failed, 1 passed, 1 error in 0.23s =============

```

In the junit-xml, the test shows up only once.
```xml
<?xml version="1.0" ?>
<testsuites>
▸       <testsuite name="pytest" errors="1" failures="1" skipped="0" tests="3" time="0.245" timestamp="2022-05-02T17:54:49.261096"             hostname="db3">
▸       ▸       <testcase classname="test_double_outcome" name="test_fail" time="0.002">
▸       ▸       ▸       <failure message="assert False">def test_fail():
&gt;       assert False
E       assert False

test_double_outcome.py:2: AssertionError</failure>
▸       ▸       </testcase>
▸       ▸       <testcase classname="test_double_outcome" name="test_pass" time="0.001">
▸       ▸       ▸       <error message="failed on teardown with &quot;assert 1 == 0&quot;">def teardown():
&gt;       assert 1 == 0
E       assert 1 == 0

test_double_outcome.py:8: AssertionError</error>
▸       ▸       </testcase>
▸       </testsuite>
</testsuites>
```

## Example 2: Last test case fails

If we switch the order of the test cases, the double reporting also shows up in the junit-xml.

```python
def test_pass():
    pass

def test_fail():
    assert False

def teardown():
    assert 1 == 0
```

```bash
> python3 -m pytest -v --junit-xml junit.xml test_double_outcome.py
===================== test session starts =====================
platform linux -- Python 3.9.2, pytest-7.1.2, pluggy-1.0.0 -- REMOVED/venv/bin/python3
cachedir: .pytest_cache
rootdir: REMOVED
collected 2 items                                             

test_double_outcome.py::test_pass PASSED                [ 50%]
test_double_outcome.py::test_fail FAILED                [100%]
test_double_outcome.py::test_fail ERROR                 [100%]

=========================== ERRORS ============================
_______________ ERROR at teardown of test_fail ________________

    def teardown():
>       assert 1 == 0
E       assert 1 == 0

test_double_outcome.py:8: AssertionError
========================== FAILURES ===========================
__________________________ test_fail __________________________

    def test_fail():
>       assert False
E       assert False

test_double_outcome.py:5: AssertionError
- generated xml file: REMOVED/junit.xml -
=================== short test summary info ===================
FAILED test_double_outcome.py::test_fail - assert False
ERROR test_double_outcome.py::test_fail - assert 1 == 0
============ 1 failed, 1 passed, 1 error in 0.25s =============
```

```xml
<?xml version="1.0" ?>
<testsuites>
▸       <testsuite name="pytest" errors="1" failures="1" skipped="0" tests="2" time="0.044" timestamp="2022-05-02T17:52:52.371978"             hostname="db3">
▸       ▸       <testcase classname="test_double_outcome" name="test_pass" time="0.001"/>
▸       ▸       <testcase classname="test_double_outcome" name="test_fail" time="0.001">
▸       ▸       ▸       <failure message="assert False">def test_fail():
&gt;       assert False
E       assert False

test_double_outcome.py:5: AssertionError</failure>
▸       ▸       </testcase>
▸       ▸       <testcase classname="test_double_outcome" name="test_fail" time="0.000">
▸       ▸       ▸       <error message="failed on teardown with &quot;assert 1 == 0&quot;">def teardown():
&gt;       assert 1 == 0
E       assert 1 == 0

test_double_outcome.py:8: AssertionError</error>
▸       ▸       </testcase>
▸       </testsuite>
</testsuites>

```

## My setup

``` bash
> pip list          
Package       Version
------------- -------
attrs         21.4.0
iniconfig     1.1.1
packaging     21.3
pip           20.3.4
pkg-resources 0.0.0
pluggy        1.0.0
py            1.11.0
pyparsing     3.0.8
pytest        7.1.2
setuptools    44.1.1
tomli         2.0.1
wheel         0.34.2


> python3 -m pytest --version
pytest 7.1.2
```

**Comment by asottile** [comment]
teardowns are always associated with the most recent test which ran

in some cases this results in multiple statuses for a single test

everything shown in the issue is expected

**Comment by gruberma** [comment]
perfect, thanks for the quick response :-)

**Comment by RonnyPfannschmidt** [comment]
@asottile as this one is about sometimes doublse, sometimes single entries in the junit report, 
i do wonder if we should take it as a bug, but its not clear how to map this sanely to junitxml if we ensure report collapsing

**Comment by asottile** [comment]
@RonnyPfannschmidt there isn't a sometimes doubling if you look at the xml -- op was just confused by which test the teardown failure was associated with

**Comment by gruberma** [comment]
> @RonnyPfannschmidt there isn't a sometimes doubling if you look at the xml -- op was just confused by which test the teardown failure was associated with

Hmm, I think the doubling doesn't always occur. In example 1,`test_pass` only occurs once in the junit-xml, but in example 2, `test_fail` occurs twice.

**Comment by storenth** [comment]
Yeah, same issue I got! duplicate test results in case of any teardown failure. Take a look at allure example to figure out the idea! I can't mark the test properly;( So, results looks like: `38 deselected, 1 failed, 1 xpassed in 1.41s`
<img width="297" alt="allure-example" src="https://user-images.githubusercontent.com/18244131/221542658-6200df26-b849-4801-9491-7a11d4f32175.png">

**Comment by petr-balogh** [comment]
Hello @RonnyPfannschmidt , I think we hit the related issue to this (see my comment [here](https://github.com/red-hat-storage/ocs-ci/issues/8842#issuecomment-1810845603)) and then our report which we are uploading to Report Portal is missing some of the properties for duplicate case. Not sure how to W/A or fix this different way to at least have our properties reported in all of the occurrences of failures in junit XML report?  Thanks

**Comment by RonnyPfannschmidt** [comment]
@petr-balogh a workaround could be to preprocess the report to merge the items manually, I haven't investigated this any further so far

**Comment by Sillocan** [comment]
This is still an issue with the pytest 8.3.3. What's really odd is that the XML has an extra test entry versus the output of pytest

**Comment by storenth** [comment]
Looks like we need to deal with test design principles because setup and teardown are steps to instrument your system but not test itself. I suggest move the bug to Discuss section.

**Comment by RonnyPfannschmidt** [comment]
Nose support was removed and i suspect that stops the teardown 

Changing it to a autouse fixture with yield should recreate them error

**Comment by gruberma** [comment]
I just checked both examples again using the latest pytest version (`8.3.5`) and the issue did indeed disappear.

I also tried it with an autouse fixture, as @RonnyPfannschmidt suggested, but I wasn't able to reproduce it yet. @RonnyPfannschmidt could you please give me feedback if my code matches the scenario you had in mind?

```python
import pytest

@pytest.fixture(autouse=True)
def erroring_fixture():
    assert 1 == 0

def test_pass():
    pass

def test_fail():
    assert False
```

**Comment by RonnyPfannschmidt** [comment]
Your code matches what i had in mind

**Comment by gruberma** [comment]
Thanks for the feedback @RonnyPfannschmidt 

Then it seems to me like the issue got resolved. Should I close it?

**Comment by undercoverturbo** [comment]
I have tested this with `8.4.1` and I am stil getting two states reported for the same test case.

This is the output I am getting:
```
demo\settings\settings_test.py::Test_Settings::test_display PASSED       [100%]
demo\settings\settings_test.py::Test_Settings::test_display ERROR        [100%]

=================================== ERRORS ====================================
_______________ ERROR at teardown of Test_Settings.test_display _______________
Test Case reported NOK during execution, look at the report for more details!
=================================== PASSES ====================================
_________________________ Test_Settings.test_display __________________________
--- generated xml file: D:\projects\[...]\pytest_junit.xml ---
=========================== short test summary info ===========================
PASSED demo\settings\settings_test.py::Test_Settings::test_display
ERROR demo\settings\settings_test.py::Test_Settings::test_display - Failed: Test Case reported NOK during execution, look at the report for more details!
=================== 1 passed, 1 error in 139.16s (0:02:19) ====================
```

My test cases track the state of test steps and only at the end it is evaluated, if the test case only contains OK test steps, if a test step is NOK, the test case fails.
I don't want to have to write the check for that in every test case, so I have it in the conftest.py file:
```
@pytest.fixture(scope="function", autouse=True)
def base_func_fixture(record_property):
    [...]
    yield

    check_test_case_status()
```

So this issue still seems to be there.

**Comment by nicoddemus** [comment]
@undercoverturbo I suspect your issue is the same "theme" but a different cause. Could you post a separate issue, along with a MWE? Thanks!

I'm closing this one then, thanks folks!
